### ISSUE TITLE
Return filename in Figure.savefig

### ISSUE THREAD

**Issue by teto** [matplotlib/matplotlib#7402]
To help us understand and resolve your issue please check that you have provided
the information below.

- [x] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...) 1.5.3
- [x] How did you install Matplotlib and Python (pip, anaconda, from source ...) pip

As per the documentation, when figure.savefig is called with a filename without extension and format is None, the function will append the extension of the format (e.g. ".png") to the filename.
Is there a way to disable that or change the prototype so that it returns the filename ?

**Comment by tacaswell** [comment]
Couldn't you just pass the format you want (and then we do not change the file name)?

**Comment by teto** [comment]
I use matplotlib savefig where the user can either choose a filename or let the default one (I configured as "output.png"). 
If the user chooses not to pass an extension in the filename, then why matplotlib would decide to change the filename in the first place ? seems like unnecessary work.
It's no big deal as it's not a wide audience program, for now I will default everything to png.

**Comment by tacaswell** [comment]
This behavior (tacking on an extension) has been there sense at least 2007.  The default format it picks is deferred to the backend you are using not hard-coded to png, thus you do not in general know _what_ format will be.  I think the code paths through this amount to the following cases:

 - If the user passes in a file name with an extension we try to infer the format from the extension (and if we fail the save fails). 
 - if the user does not pass a format or a files name with an extension, we append the format that is written out
 - If the user passes in a file handle we just write to it (which is a path you could take go get the desired behavior).  In this case there is no file name so it is unclear what should be done for the return value if we started to return the file name (`None` probably?).
 - There is an open PR which special cases passing in `None` as the first argument to mean 'return the image as a `Btyes` object' https://github.com/matplotlib/matplotlib/pull/6909 which I think should have priority over this this request and `Bytes` or `str` or `None` as the return type is a bit too mixed I think.

In conclusion, while this is a reasonable idea and probably closer to how it would be written if starting from scratch _now_, I think we have too much other stuff going on with this function to implement this on top.  Sorry :(
